dbms/src/Flash/Pipeline/Exec/PipelineExec.cpp

```diff
@@ -0,0 +1,110 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Exec/PipelineExec.h>
+
+namespace DB
+{
+#define CHECK_IS_CANCELLED(exec_status)        \
+    if (unlikely((exec_status).isCancelled())) \
+        return OperatorStatus::CANCELLED;
+
+/**
+ *  sink_op   transform_op    ...   transform_op   source_op
+ *
+ *  prepare────►tryOutput───► ... ───►tryOutput────►read────┐
+ *                                                          │ block
+ *    write◄────transform◄─── ... ◄───transform◄────────────┘
+ */
+OperatorStatus PipelineExec::execute(PipelineExecutorStatus & exec_status)
+{
+    Block block;
+    size_t start_transform_op_index = 0;
+    auto op_status = fetchBlock(block, start_transform_op_index, exec_status);
+    // If the state `fetchBlock` returns isn't `HAS_OUTPUT`, it means that `fetchBlock` did not return a block and is an exception state.
+    if (op_status != OperatorStatus::HAS_OUTPUT)
+        return op_status;
+
+    // start from the next transform after fetched block transform.
+    for (size_t transform_op_index = start_transform_op_index; transform_op_index < transform_ops.size(); ++transform_op_index)
+    {
+        CHECK_IS_CANCELLED(exec_status);
+        auto op_status = transform_ops[transform_op_index]->transform(block);
+        if (op_status != OperatorStatus::HAS_OUTPUT)
+            return prepareSpillOpIfNeed(op_status, transform_ops[transform_op_index]);
+    }
+    CHECK_IS_CANCELLED(exec_status);
+    return prepareSpillOpIfNeed(sink_op->write(std::move(block)), sink_op);
+}
+
+// try fetch block from transform_ops and source_op.
+OperatorStatus PipelineExec::fetchBlock(
+    Block & block,
+    size_t & start_transform_op_index,
+    PipelineExecutorStatus & exec_status)
+{
+    CHECK_IS_CANCELLED(exec_status);
+    auto op_status = sink_op->prepare();
+    if (op_status != OperatorStatus::NEED_INPUT)
+        return prepareSpillOpIfNeed(op_status, sink_op);
+    for (int64_t index = transform_ops.size() - 1; index >= 0; --index)
+    {
+        CHECK_IS_CANCELLED(exec_status);
+        auto op_status = transform_ops[index]->tryOutput(block);
+        if (op_status != OperatorStatus::NEED_INPUT)
+        {
+            // Once the transform op tryOutput has succeeded, execution will begin with the next transform op.
+            start_transform_op_index = index + 1;
+            return prepareSpillOpIfNeed(op_status, transform_ops[index]);
+        }
+    }
+    CHECK_IS_CANCELLED(exec_status);
+    start_transform_op_index = 0;
+    return prepareSpillOpIfNeed(source_op->read(block), source_op);
+}
+
+OperatorStatus PipelineExec::await(PipelineExecutorStatus & exec_status)
+{
+    CHECK_IS_CANCELLED(exec_status);
+
+    auto op_status = sink_op->await();
+    if (op_status != OperatorStatus::NEED_INPUT)
+        return op_status;
```

Comment:
Shall we assert `op_status == OperatorStatus::WAITING`?

Reply:
no, finish status, spilling status and waiting status can be returned in any method.